src/Perspex.Controls/Templates/DataTemplateExtensions.cs

```diff
@@ -60,9 +68,11 @@ public static IControl MaterializeDataTemplate(this IControl control, object dat
         /// <returns>The data template or null if no matching data template was found.</returns>
         public static IDataTemplate FindDataTemplate(this IControl control, object data)
         {
+            var templatesMaterializedFrom = control.GetSelfAndLogicalAncestors().OfType<IControl>()
+                                                    .Select(c => c.MaterializedFrom).Where(dt => dt != null);
             foreach (var i in control.GetSelfAndLogicalAncestors().OfType<IControl>())
             {
-                foreach (IDataTemplate dt in i.DataTemplates)
+                foreach (IDataTemplate dt in i.DataTemplates.Except(templatesMaterializedFrom))
```

Comment:
This will cause the whole logical tree to be traversed for each control in the logical tree. i.e. it's introducing logarithmic complexity.